adoc/chapters/programming_interface.adoc

```diff
@@ -19401,481 +19401,697 @@ that all of the built-in functions fulfill the same requirements for both
+These are not actual type names available to SYCL applications.  Instead, they
+are a shorthand used in this document to describe the set of overloads for
+these functions.  For example the [code]#fmax# function is specified as:
 
+[source]
+----
+genfloat fmax(genfloat x, genfloat y)
+----
+
+The generic type [code]#genfloat# is a stand-in for a set of a actual type
+names: [code]#float#, [code]#double#, etc.  To construct the actual overloads
+that are available, replace the generic type name with the name of each of
+these actual types:
+
+[source]
+----
+float fmax(float x, float y)
+double fmax(double x, double y)
+template<int N> vec<float,N> fmax(vec<float,N> x, vec<float,N> y)
+template<size_t M> marray<float,M> fmax(marray<float,M> x, marray<float,M> y)
+// etc.
+----
+
+When determining the set of overloads that are available, use the following
+rules:
+
+* If the specification of a function has several occurrences of the same
+  generic type name, replace each occurrence with the same actual type.  For
+  example, the [code]#fmax# example above has three occurrences of
+  [code]#genfloat#: one for the return type and one for each parameter type.
+  The available overloads include only the cases where both parameters have
+  the same type, and the return type always matches the type of the parameters.
+
+* Some of the generic type names can be replaced with a templated type, where
+  the identifiers [code]#N#, [code]#M#, [code]#Space#, and [code]#IsDecorated#
+  represent template parameters.  When this happens, the overload is a
+  templated function, as shown above with the [code]#fmax# example.  Only a few
+  combinations of template parameters are possible, and the following list
+  tells the order and type of the template parameters for each possible
+  combination:
+
+** If there is a single template parameter [code]#N# (number of dimensions of
+   the [code]#vec# type):
++
+[source]
+----
+template<int N>
+----
+
+** If there is a single template parameter [code]#M# (number of dimensions of
+   the [code]#marray# type):
++
+[source]
+----
+template<size_t M>
+----
+
+** If there are two template parameters [code]#Space# and [code]#IsDecorated#
+   (address space and decoration for the [code]#multi_ptr# type):
++
+[source]
+----
+template<access::address_space Space, access::decorated IsDecorated>
+----
 
+** If there are three template parameters [code]#N#, [code]#Space# and
+   [code]#IsDecorated#:
++
+[source]
+----
+template<int N, access::address_space Space, access::decorated IsDecorated>
+----
+
+** If there are three template parameters [code]#M#, [code]#Space# and
+   [code]#IsDecorated#:
++
+[source]
+----
+template<size_t M, access::address_space Space, access::decorated IsDecorated>
+----
+
+* Unless specified otherwise, the template parameter [code]#N# is constrained
+  to the values 2, 3, 4, 8, or 16.
+
+* Unless specified otherwise, the template parameter [code]#M# is constrained
+  to a positive value.
```

Comment:
Is this required?  If `M` is `size_t`, isn't it always unsigned?

Reply:
This prevents the value from being zero.

Thinking about this more, though ... maybe we don't need either of these constraints on `N` or `M`.  Presumably, the types `vec` and `marray` should have some sort of `static_assert` that their template parameter is in range.  In that case, there is no need to redundantly put the same constraint on these functions.

Reply:
Oh, I see.  If you do keep these constraints, I think "M must be greater than zero" is a clearer way to say that.

But I'm in favor of removing the duplicated constraints.

Reply:
I removed the constraints for `N` and `M` in 1133eefba7b155f68320b0776417ed9afa337040.